File: Day-3/piece-2/Models/Quote.cs
using QuotesApi.Exceptions;

namespace QuotesApi.Models;

public class Quote
{
    public int Id { get; private set; }
    public string Author { get; private set; } = string.Empty;
    public string Text { get; private set; } = string.Empty;
    public DateTime CreatedAt { get; private set; }
    public bool IsDeleted { get; private set; }
    public int? OwnerId { get; private set; }

    private Quote() { } // For EF Core

    public static Quote Create(string author, string text, int? ownerId = null)
    {
        if (string.IsNullOrWhiteSpace(author) || author.Length > 200)
            throw new DomainException("Author must be between 1 and 200 characters.");

        if (string.IsNullOrWhiteSpace(text) || text.Length > 1000)
            throw new DomainException("Text must be between 1 and 1000 characters.");

        return new Quote
        {
            Author = author.Trim(),
            Text = text.Trim(),
            CreatedAt = DateTime.UtcNow,
            OwnerId = ownerId
        };
    }

    public void SoftDelete()
    {
        IsDeleted = true;
    }
}

File: Day-22/piece-1/src/app/models/quote.ts
// Domain models — mirror EXACTLY what the Week-1 API returns.
//
//   GET /api/quotes?page={n}&size={n}   →  Quote[]   (list; !IsDeleted, paged)
//   GET /api/quotes/{id}                →  Quote     (detail) or 404
//
// The JSON is the serialised EF entity (QuotesApi/Models/Quote.cs), so the
// field names here must match its public properties one-for-one. No `any`
// anywhere — list and detail share this single typed shape.

export interface Quote {
  id:        number;
  author:    string;
  text:      string;        // NOT "body"/"content" — the entity property is Text
  createdAt: string;        // ISO-8601 (DateTime serialised)
  isDeleted: boolean;       // present on the entity; API already filters these out
  ownerId:   number | null; // int? on the server → number | null here
}

// Request body for POST /api/quotes (QuotesApi/DTOs/CreateQuoteRequest.cs).
// ONLY these two fields exist on the contract — Author [Required, MaxLength(200)]
// and Text [Required, MaxLength(1000)]. ownerId is taken from the JWT on the
// server, NOT sent by the client; id/createdAt/isDeleted are server-assigned.
export interface CreateQuoteRequest {
  author: string;
  text:   string;
}
